cpp/src/arrow/device.h

```diff
@@ -98,6 +100,54 @@ class ARROW_EXPORT Device : public std::enable_shared_from_this<Device>,
   /// \brief Return the DeviceAllocationType of this device
   virtual DeviceAllocationType device_type() const = 0;
 
+  class SyncEvent;
+
+  /// \brief EXPERIMENTAL: An opaque wrapper for Device-specific streams
+  ///
+  /// In essence this is just a wrapper around a void* to represent the
+  /// standard concept of a stream/queue on a device. Derived classes
+  /// should be trivially constructible from it's device-specific counterparts.
+  class ARROW_EXPORT Stream {
```

Comment:
I think we want to call this `DeviceStream` otherwise there's some latent confusion with `InputStream` and potentially other concepts?

Reply:
Even though it's a subclass of `Device`? I think it would be redundant to be `Device::DeviceStream`?

Reply:
Woops, sorry. I had entirely missed the subclass aspect. Nevermind :-)